File: python/src/websocket/state/price.py
"""Price history state management."""

from dataclasses import dataclass
from typing import Optional

from ..types import Candle, PriceHistoryData
from ...shared.types import Resolution


@dataclass
class PriceHistoryKey:
    """Key for price history subscriptions."""

    orderbook_id: str
    resolution: str

    def __hash__(self):
        return hash((self.orderbook_id, self.resolution))

    def __eq__(self, other):
        if not isinstance(other, PriceHistoryKey):
            return False
        return (
            self.orderbook_id == other.orderbook_id
            and self.resolution == other.resolution
        )


class PriceHistory:
    """Price history state for a single orderbook/resolution pair."""

    def __init__(self, orderbook_id: str, resolution: str, include_ohlcv: bool = False):
        self.orderbook_id = orderbook_id
        self.resolution = resolution
        self.include_ohlcv = include_ohlcv
        self._candles: list[Candle] = []  # Sorted by timestamp (newest first)
        self._candle_index: dict[int, int] = {}  # timestamp -> index
        self._last_timestamp: Optional[int] = None
        self._server_time: Optional[int] = None
        self._has_snapshot: bool = False

    def apply_snapshot(self, data: PriceHistoryData) -> None:
        """Apply a snapshot (historical candles)."""
        self._candles.clear()
        self._candle_index.clear()

        # Apply candles (they come newest-first from server)
        for candle in data.prices:
            idx = len(self._candles)
            self._candle_index[candle.t] = idx
            self._candles.append(candle)

        self._last_timestamp = data.last_timestamp
        self._server_time = data.server_time
        self._has_snapshot = True

        if data.include_ohlcv is not None:
            self.include_ohlcv = data.include_ohlcv

    def apply_update(self, data: PriceHistoryData) -> None:
        """Apply an update (new or updated candle)."""
        candle = data.to_candle()
        if candle:
            self._update_or_append_candle(candle)

    def _update_or_append_candle(self, candle: Candle) -> None:
        """Update an existing candle or append a new one."""
        if candle.t in self._candle_index:
            # Update existing candle
            idx = self._candle_index[candle.t]
            self._candles[idx] = candle
        else:
            # New candle - insert at the correct position (newest first)
            insert_pos = 0
            for i, c in enumerate(self._candles):
                if c.t < candle.t:
                    insert_pos = i
                    break
            else:
                insert_pos = len(self._candles)

            # Update indices for candles that will shift
            new_index = {}
            for ts, idx in self._candle_index.items():
                if idx >= insert_pos:
                    new_index[ts] = idx + 1
                else:
                    new_index[ts] = idx
            new_index[candle.t] = insert_pos
            self._candle_index = new_index

            self._candles.insert(insert_pos, candle)

            # Trim to max 1000 candles
            while len(self._candles) > 1000:
                removed = self._candles.pop()
                self._candle_index.pop(removed.t, None)

        # Update last timestamp
        if self._candles:
            self._last_timestamp = self._candles[0].t

    def apply_heartbeat(self, data: PriceHistoryData) -> None:
        """Handle heartbeat (update server time)."""
        self._server_time = data.server_time

    def apply_event(self, data: PriceHistoryData) -> None:
        """Apply any price history event."""
        if data.event_type == "snapshot":
            self.apply_snapshot(data)
        elif data.event_type == "update":
            self.apply_update(data)
        elif data.event_type == "heartbeat":
            self.apply_heartbeat(data)

    def candles(self) -> list[Candle]:
        """Get all candles (newest first)."""
        return self._candles.copy()

    def recent_candles(self, n: int) -> list[Candle]:
        """Get the N most recent candles."""
        return self._candles[:n]

    def get_candle(self, timestamp: int) -> Optional[Candle]:
        """Get a candle by timestamp."""
        idx = self._candle_index.get(timestamp)
        if idx is not None:
            return self._candles[idx]
        return None

    def latest_candle(self) -> Optional[Candle]:
        """Get the most recent candle."""
        return self._candles[0] if self._candles else None

    def oldest_candle(self) -> Optional[Candle]:
        """Get the oldest candle."""
        return self._candles[-1] if self._candles else None

    def current_midpoint(self) -> Optional[str]:
        """Get current midpoint price (from most recent candle)."""
        if self._candles:
            return self._candles[0].m
        return None

    def current_best_bid(self) -> Optional[str]:
        """Get current best bid (from most recent candle)."""
        if self._candles:
            return self._candles[0].bb
        return None

    def current_best_ask(self) -> Optional[str]:
        """Get current best ask (from most recent candle)."""
        if self._candles:
            return self._candles[0].ba
        return None

    def candle_count(self) -> int:
        """Number of candles."""
        return len(self._candles)

    def has_snapshot(self) -> bool:
        """Whether the price history has received its initial snapshot."""
        return self._has_snapshot

    def last_timestamp(self) -> Optional[int]:
        """Last candle timestamp."""
        return self._last_timestamp

    def server_time(self) -> Optional[int]:
        """Server time from last message."""
        return self._server_time

    def resolution_enum(self) -> Optional[Resolution]:
        """Get resolution as enum."""
        try:
            return Resolution.from_str(self.resolution)
        except ValueError:
            return None

    def clear(self) -> None:
        """Clear the price history (for disconnect/resync)."""
        self._candles.clear()
        self._candle_index.clear()
        self._last_timestamp = None
        self._server_time = None
        self._has_snapshot = False

File: typescript/src/websocket/state/price.ts
/**
 * Price history state management.
 *
 * Maintains local state for price history candles.
 */

import { Resolution } from "../../shared";
import type { Candle, PriceHistoryData } from "../types";
import { toCandle } from "../types";

/**
 * Key for price history subscriptions.
 */
export class PriceHistoryKey {
  readonly orderbookId: string;
  readonly resolution: string;

  constructor(orderbookId: string, resolution: string) {
    this.orderbookId = orderbookId;
    this.resolution = resolution;
  }

  toString(): string {
    return `${this.orderbookId}:${this.resolution}`;
  }

  equals(other: PriceHistoryKey): boolean {
    return (
      this.orderbookId === other.orderbookId &&
      this.resolution === other.resolution
    );
  }
}

/**
 * Price history state for a single orderbook/resolution pair.
 */
export class PriceHistory {
  /** Maximum number of candles to maintain */
  private static readonly MAX_CANDLES = 1000;

  /** Orderbook identifier */
  readonly orderbookId: string;
  /** Resolution (1m, 5m, 15m, 1h, 4h, 1d) */
  readonly resolution: string;
  /** Whether OHLCV data is included */
  includeOhlcv: boolean;
  /** Candles sorted by timestamp (newest first) */
  private _candles: Candle[] = [];
  /** Index by timestamp for fast lookup */
  private candleIndex: Map<number, number> = new Map();
  /** Last server timestamp */
  private _lastTimestamp?: number;
  /** Server time from last message */
  private _serverTime?: number;
  /** Whether initial snapshot has been received */
  private _hasSnapshot: boolean = false;

  constructor(orderbookId: string, resolution: string, includeOhlcv: boolean) {
    this.orderbookId = orderbookId;
    this.resolution = resolution;
    this.includeOhlcv = includeOhlcv;
  }

  /**
   * Apply a snapshot (historical candles).
   */
  applySnapshot(data: PriceHistoryData): void {
    // Clear existing state
    this._candles = [];
    this.candleIndex.clear();

    // Apply candles (they come newest-first from server)
    for (const candle of data.prices) {
      const idx = this._candles.length;
      this.candleIndex.set(candle.t, idx);
      this._candles.push(candle);
    }

    this._lastTimestamp = data.last_timestamp;
    this._serverTime = data.server_time;
    this._hasSnapshot = true;

    // Update include_ohlcv if provided
    if (data.include_ohlcv !== undefined) {
      this.includeOhlcv = data.include_ohlcv;
    }
  }

  /**
   * Apply an update (new or updated candle).
   */
  applyUpdate(data: PriceHistoryData): void {
    const candle = toCandle(data);
    if (candle) {
      this.updateOrAppendCandle(candle);
    }
  }

  /**
   * Update an existing candle or append a new one.
   */
  private updateOrAppendCandle(candle: Candle): void {
    const existingIdx = this.candleIndex.get(candle.t);
    if (existingIdx !== undefined) {
      // Update existing candle
      this._candles[existingIdx] = candle;
    } else {
      // New candle - insert at the correct position (newest first)
      const insertPos = this._candles.findIndex((c) => c.t < candle.t);
      const pos = insertPos === -1 ? this._candles.length : insertPos;

      // Update indices for candles that will shift
      for (const [ts, idx] of this.candleIndex.entries()) {
        if (idx >= pos) {
          this.candleIndex.set(ts, idx + 1);
        }
      }

      this.candleIndex.set(candle.t, pos);
      this._candles.splice(pos, 0, candle);

      // Trim to max candles limit
      while (this._candles.length > PriceHistory.MAX_CANDLES) {
        const removed = this._candles.pop();
        if (removed) {
          this.candleIndex.delete(removed.t);
        }
      }
    }

    // Update last timestamp
    if (this._candles.length > 0) {
      this._lastTimestamp = this._candles[0].t;
    }
  }

  /**
   * Handle heartbeat (update server time).
   */
  applyHeartbeat(data: PriceHistoryData): void {
    this._serverTime = data.server_time;
  }

  /**
   * Apply any price history event.
   */
  applyEvent(data: PriceHistoryData): void {
    switch (data.event_type) {
      case "snapshot":
        this.applySnapshot(data);
        break;
      case "update":
        this.applyUpdate(data);
        break;
      case "heartbeat":
        this.applyHeartbeat(data);
        break;
      default:
        console.warn(`Unknown price history event type: ${data.event_type}`);
    }
  }

  /**
   * Get all candles (newest first).
   */
  candles(): Candle[] {
    return this._candles;
  }

  /**
   * Get the N most recent candles.
   */
  recentCandles(n: number): Candle[] {
    return this._candles.slice(0, Math.min(n, this._candles.length));
  }

  /**
   * Get a candle by timestamp.
   */
  getCandle(timestamp: number): Candle | undefined {
    const idx = this.candleIndex.get(timestamp);
    if (idx === undefined) return undefined;
    return this._candles[idx];
  }

  /**
   * Get the most recent candle.
   */
  latestCandle(): Candle | undefined {
    return this._candles[0];
  }

  /**
   * Get the oldest candle.
   */
  oldestCandle(): Candle | undefined {
    return this._candles[this._candles.length - 1];
  }

  /**
   * Get current midpoint price (from most recent candle).
   */
  currentMidpoint(): string | undefined {
    return this._candles[0]?.m;
  }

  /**
   * Get current best bid (from most recent candle).
   */
  currentBestBid(): string | undefined {
    return this._candles[0]?.bb;
  }

  /**
   * Get current best ask (from most recent candle).
   */
  currentBestAsk(): string | undefined {
    return this._candles[0]?.ba;
  }

  /** Number of candles */
  candleCount(): number {
    return this._candles.length;
  }

  /** Whether the price history has received its initial snapshot */
  hasSnapshot(): boolean {
    return this._hasSnapshot;
  }

  /** Last candle timestamp */
  lastTimestamp(): number | undefined {
    return this._lastTimestamp;
  }

  /** Server time from last message */
  serverTime(): number | undefined {
    return this._serverTime;
  }

  /**
   * Get resolution as enum.
   */
  resolutionEnum(): Resolution | undefined {
    switch (this.resolution) {
      case "1m":
        return Resolution.OneMinute;
      case "5m":
        return Resolution.FiveMinutes;
      case "15m":
        return Resolution.FifteenMinutes;
      case "1h":
        return Resolution.OneHour;
      case "4h":
        return Resolution.FourHours;
      case "1d":
        return Resolution.OneDay;
      default:
        return undefined;
    }
  }

  /** Clear the price history (for disconnect/resync) */
  clear(): void {
    this._candles = [];
    this.candleIndex.clear();
    this._lastTimestamp = undefined;
    this._serverTime = undefined;
    this._hasSnapshot = false;
  }
}
